cpp/src/arrow/record_batch.cc

```diff
@@ -78,6 +79,52 @@ class SimpleRecordBatch : public RecordBatch {
 
   std::shared_ptr<ArrayData> column_data(int i) const override { return columns_[i]; }
 
+  Status AddColumn(int i, const std::shared_ptr<Field>& field,
+                   const std::shared_ptr<ArrayData>& column,
+                   std::shared_ptr<RecordBatch>* out) const override {
+    if (i < 0 || i > num_columns() + 1) {
+      return Status::Invalid("Invalid column index");
+    }
+    if (field == nullptr) {
+      std::stringstream ss;
+      ss << "Field " << i << " was null";
+      return Status::Invalid(ss.str());
+    }
+    if (column == nullptr) {
+      std::stringstream ss;
+      ss << "Column " << i << " was null";
+      return Status::Invalid(ss.str());
+    }
+    if (!field->type()->Equals(column->type)) {
+      std::stringstream ss;
+      ss << "Column data type " << field->type()->name()
+         << " does not match field data type " << column->type->name();
+      return Status::Invalid(ss.str());
+    }
+    if (column->length != num_rows_) {
+      std::stringstream ss;
+      ss << "Added column's length must match record batch's length. Expected length "
+         << num_rows_ << " but got length " << column->length;
+      return Status::Invalid(ss.str());
+    }
+
+    std::shared_ptr<Schema> new_schema;
+    RETURN_NOT_OK(schema_->AddField(i, field, &new_schema));
```

Comment:
We could leave the boundschecking above to `Schema::AddField` -- could you also check whether that function has the issues described above?